src/main/kotlin/io/sc3/peripherals/posters/PosterPrintData.kt

```diff
@@ -0,0 +1,94 @@
+package io.sc3.peripherals.posters
+
+import io.sc3.library.ext.optString
+import io.sc3.library.ext.putOptString
+import io.sc3.peripherals.config.ScPeripheralsConfig.config
+import io.sc3.peripherals.mixin.MapColorAccessor
+import io.sc3.peripherals.posters.PosterItem.Companion.POSTER_KEY
+import net.minecraft.block.MapColor
+import net.minecraft.nbt.NbtCompound
+import net.minecraft.text.Text
+import kotlin.math.roundToInt
+import kotlin.math.sign
+
+const val MAX_LABEL_LENGTH = 48
+const val MAX_TOOLTIP_LENGTH = 256
+
+fun getDefaultPalette() = MapColorAccessor.getColors().map { it?.color ?: MapColor.CLEAR.color }.toIntArray()
+
+data class PosterPrintData(
+  private val initialLabel: String? = null,
+  var tooltip: String? = null,
+  val colors: ByteArray = ByteArray(128 * 128) { 0 },
+  var palette: IntArray = getDefaultPalette(),
+  var posterId: String? = null,
+) {
+  var label: String? = initialLabel
+    set(value) {
+      field = value?.takeIf { it.isValidLabel() }
+      labelText = field?.let { Text.of(it) }
+    }
+
+  var labelText: Text? = initialLabel?.let { Text.of(it) }
+    private set
+
+  fun computeCosts(): Int {
+    val pixels = colors.count { it != 0.toByte() }
+    return (posterInkCost*(pixels / 16384.0)).roundToInt().coerceAtLeast(pixels.sign)
+  }
+
+  fun toNbt(): NbtCompound {
+    val nbt = toItemNbt()
+    nbt.putByteArray("colors", colors)
+    nbt.putIntArray("palette", palette)
+    return nbt
+  }
+
+  fun toItemNbt(): NbtCompound {
+    val nbt = NbtCompound()
+    nbt.putOptString("label", label?.takeIf { it.isValidLabel() })
+    nbt.putOptString("tooltip", tooltip?.takeIf { it.isValidTooltip() })
+    posterId?.let { nbt.putString(POSTER_KEY, it) }
```

Comment:
This is pretty much just what putOptString does

```suggestion
    nbt.putOptString(POSTER_KEY, posterId)
```